test/e2e/client_tls_test.go

```diff
@@ -384,6 +313,535 @@ func TestClientTLS(t *testing.T) {
+			clientPod.Spec.Containers[0].VolumeMounts = []corev1.VolumeMount{{
+				Name:      "client-certificates",
+				MountPath: "/tmp/tls/",
+				ReadOnly:  true,
+			}}
+			clientPodName := types.NamespacedName{
+				Name:      clientPod.Name,
+				Namespace: clientPod.Namespace,
+			}
+			if err := kclient.Create(context.TODO(), clientPod); err != nil {
+				t.Fatalf("failed to create pod %q: %v", clientPodName, err)
+			}
+			defer assertDeleted(t, kclient, clientPod)
+
+			err = wait.PollImmediate(2*time.Second, 3*time.Minute, func() (bool, error) {
+				if err := kclient.Get(context.TODO(), clientPodName, clientPod); err != nil {
+					t.Logf("failed to get client pod %q: %v", clientPodName, err)
+					return false, nil
+				}
+				for _, cond := range clientPod.Status.Conditions {
+					if cond.Type == corev1.PodReady {
+						return cond.Status == corev1.ConditionTrue, nil
+					}
+				}
+				return false, nil
+			})
+			if err != nil {
+				t.Fatalf("timed out waiting for pod %q to become ready: %v", clientPodName, err)
+			}
+
+			// We need an IP address to which to send requests.  The test client
+			// runs inside the cluster, so we can use the custom router's internal
+			// service address.
+			service := &corev1.Service{}
+			serviceName := controller.InternalIngressControllerServiceName(ic)
+			if err := kclient.Get(context.TODO(), serviceName, service); err != nil {
+				t.Fatalf("failed to get service %q: %v", serviceName, err)
+			}
+
+			// We need some route to which to send requests.  The canary route
+			// should always exist and be responsive, so use that.
+			route := &routev1.Route{}
+			routeName := controller.CanaryRouteName()
+			err = wait.PollImmediate(1*time.Second, 1*time.Minute, func() (bool, error) {
+				if err := kclient.Get(context.TODO(), routeName, route); err != nil {
+					t.Logf("failed to get route %q: %v", routeName, err)
+					return false, nil
+				}
+				return true, nil
+			})
+			if err != nil {
+				t.Fatalf("failed to observe route %q: %v", routeName, err)
+			}
+
+			for certName := range tcCerts.ClientCerts.Accepted {
+				if _, err := curl(t, clientPod, certName, route.Spec.Host, service.Spec.ClusterIP, false); err != nil {
+					t.Errorf("Failed to curl route with cert %q: %v", certName, err)
+					crlString := ""
+					for _, crl := range tcCerts.CRLs {
+						crlString += crl
+					}
+					t.Errorf("CA Bundle:\n%s\nClient Certificate:\n%s\nCRLs:\n%s", tcCerts.CABundle, tcCerts.ClientCerts.Accepted[certName].CertPem, crlString)
+				}
+			}
+			for certName := range tcCerts.ClientCerts.Rejected {
+				if httpStatusCode, err := curl(t, clientPod, certName, route.Spec.Host, service.Spec.ClusterIP, false); err != nil {
+					if httpStatusCode == 0 {
+						// TLS/SSL verification failures result in a 0 http status code (no connection is made to the backend, so no http status code is returned).
+						continue
+					}
+					t.Errorf("Unexpected error from curl for cert %q: %v", certName, err)
```

Comment:
Would it be beneficial to add logging of the `httpStatusCode` for future debugging?